// RUN: %target-swift-frontend %s -parse -verify

// Distributed under the terms of the MIT license
// Test case submitted to project by https://github.com/practicalswift (practicalswift)
// Test case found by fuzzing

func f<c>(f.c { // expected-error{{}} expected-error{{}} expected-error{{}} expected-error{{}}
typealias f == h>(T][], e: C(f("a: AnyObject, g : B // expected-error{{}}
println(c) -> : c: c) { // expected-note{{}}
func ^(A> Any, end: a { // expected-error{{}}
}
b) { // expected-error{{}} expected-error{{}}
}
let c: c = c(): // expected-error{{}} expected-error{{}} expected-error{{}}
t
// expected-error{{}}